python_modules/dagster/dagster/_core/definitions/asset_graph.py

```
@@ -1,56 +1,122 @@
+                group_names_by_key.update(asset.group_names_by_key)
             else:
                 check.failed(f"Expected SourceAsset or AssetsDefinition, got {type(asset)}")
+        return AssetGraph(
+            asset_dep_graph=generate_asset_dep_graph(assets_defs, source_assets),
+            source_asset_keys={source_asset.key for source_asset in source_assets},
+            partitions_defs_by_key=partitions_defs_by_key,
+            partition_mappings_by_key=partition_mappings_by_key,
+            group_names_by_key=group_names_by_key,
+        )
 
-        self.assets_defs = assets_defs
-        self.all_assets_by_key = {
-            asset_key: assets_def for assets_def in assets_defs for asset_key in assets_def.keys
-        }
-
-        self.asset_dep_graph = generate_asset_dep_graph(assets_defs, source_assets)
-        self.source_asset_keys = {source_asset.key for source_asset in source_assets}
+    @staticmethod
+    def from_external_assets(external_asset_nodes: Sequence["ExternalAssetNode"]) -> "AssetGraph":
+        upstream = {}
+        downstream = {}
+        partitions_defs_by_key = {}
+        group_names_by_key = {}
+
+        for node in external_asset_nodes:
+            upstream[node.asset_key] = {dep.upstream_asset_key for dep in node.dependencies}
+            downstream[node.asset_key] = {dep.downstream_asset_key for dep in node.depended_by}
+            partitions_defs_by_key[node.asset_key] = (
+                node.partitions_def_data.get_partitions_definition()
+                if node.partitions_def_data
+                else None
+            )
+            group_names_by_key[node.asset_key] = node.group_name
+
+        return AssetGraph(
+            asset_dep_graph={"upstream": upstream, "downstream": downstream},
+            source_asset_keys=set(),  # TODO
+            partitions_defs_by_key=partitions_defs_by_key,
+            partition_mappings_by_key=dict(),  # TODO
```

Can we raise an error if any of the assets have partition mappings or if someone tries to use the partition mapping or something? Otherwise, I think it will be easy to write bugs that expect this to work.